### PR TITLE
Serialize the RNGs

### PR DESCRIPTION
**Pull request by trivialfis**
This PR serializes the random states used by the context and by the column sampler. The column sampler stores a state, as we need to broadcast the seed for distributed training. Please note that the configuration is not stable across releases. The `Learner` class discards the configuration automatically if the versions don't match.

*todos*
- [x] Check accuracy is not changed.
- [x] Check cross platform compatibility.
- [x] Unify the code paths for rng serialization.

Related:
- https://github.com/dmlc/xgboost/pull/12037
- https://github.com/dmlc/xgboost/issues/10324
- https://github.com/dmlc/xgboost/pull/12043

Close:
- https://github.com/dmlc/xgboost/issues/6711
- https://github.com/dmlc/xgboost/pull/10354

### CODE REVIEW

File: python-package/xgboost/testing/continuation.py [hunk 1]
```diff
@@ -57,3 +58,97 @@ def recursive_compare(obj_0: T, obj_1: T) -> None:
...
+        "objective": objective,
+        "subsample": subsample,
+        "sampling_method": sampling_method,
+        "colsample_bylevel": colsample_bylevel,
+        "seed_per_iteration": seed_per_iteration,
+        "booster": booster,
+    }
+    if num_class > 1:
+        params["num_class"] = num_class
+    if booster == "dart":
```
Comment (RAMitchell) [inline python-package/xgboost/testing/continuation.py]: Dart booster is gone.

File: src/common/random.h [hunk 2]
```diff
@@ -180,12 +185,26 @@ class ColumnSampler {
...
   collective::SafeColl(rc);
   auto cs = std::make_shared<common::ColumnSampler>(seed);
   return cs;
 }
+
+void SaveRng(Json* p_out, RandomEngine const& rng);
+void LoadRng(Json const& in, RandomEngine* rng);
+
+// The utility function handles loading configuration before the model is fit. We don't
+// want to initialize the column sampler in a constructor, to avoid creating side effect in
```
Comment (RAMitchell) [inline src/common/random.h]: Why is it bad to create a side effect in the constructor but not otherwise?

Comment (trivialfis) [reply src/common/random.h]: Think about loading a model or extracting the parameter names for validation can alter the random engine. It's quite difficult to keep the training continuation consistent if anything other than the actual training step can alter the RNG.

File: python-package/xgboost/testing/continuation.py (hunk 1 above)
Comment (trivialfis) [reply python-package/xgboost/testing/continuation.py]: I will update

**Comment by trivialfis** [comment]
> Would you consider changing the design of column sampler to require the context rng every time it takes a sample instead of maintaining its own stream? 

Let me give it a go.

File: src/common/random.h (hunk 2 above)
Comment (RAMitchell) [reply src/common/random.h]: OK as a general rule I wanted to go towards using constructors instead of the delayed initialization type pattern that's often used. I think we just need to have good tests like you have done here and be mindful of alterations to global state then constructors - I think it's always possible to create a shared pointer automatically lazy initialization situation if needed.

File: python-package/xgboost/testing/continuation.py (hunk 1 above)
Comment (trivialfis) [reply python-package/xgboost/testing/continuation.py]: Done.

**Comment by trivialfis** [comment]
@RAMitchell Removed the `rng` and `ctx` member variables from the column sampler and moved the allocation to the constructor. The column sampler `Init` call is still used lazily (unchanged) as it requires sampling parameters.
